### PR TITLE
Map - support Azure provider

### PR DESCRIPTION
**Pull request by Zedwag**

### CODE REVIEW

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js [hunk 1]
```diff
@@ -0,0 +1,884 @@
+/* global atlas */
+
+import $ from 'jquery';
+import testing from './utils.js';
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js]: [code style]
```suggestion
import { MARKERS, ROUTES } from './utils.js';
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js [hunk 2]
```diff
@@ -0,0 +1,884 @@
...
+
+                    done();
+                }
+            });
+        });
+    });
+
+    QUnit.test('map should change style on runtime type update', function(assert) {
+        const done = assert.async();
+        const d = $.Deferred();
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js]: [code style]
Let's update everywhere
```suggestion
        const mapReadyDeferred = $.Deferred();
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js [hunk 3]
```diff
@@ -0,0 +1,884 @@
...
+
+    QUnit.test('center should be geocoded if adress is passed as a string', function(assert) {
+        const done = assert.async();
+        const center = 'Cedar Park, Texas';
+
+        $('#map').dxMap({
+            provider: 'azure',
+            center,
+            onReady: () => {
+                assert.deepEqual(atlas.cameraOptions.center, this.geocodedCoordinates, 'center coordinated are correct');
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js]:
```suggestion
                assert.deepEqual(atlas.cameraOptions.center, this.geocodedCoordinates, 'center coordinates are correct');
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js [hunk 4]
```diff
@@ -0,0 +1,884 @@
...
+                }],
+                onReady: () => {
+                    assert.deepEqual(atlas.lineLayerOptions.strokeDashArray, expectedStrokeDashArray, 'strokeDashArray is correct');
+
+                    done();
+                }
+            });
+        });
+    });
+});
```
Comment (ksercs) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js:952]: [testing coverage]
- disabled state
- event subscriptions (move/click)

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+import DynamicProvider from './m_provider.dynamic';
+
+const window = getWindow();
+
+declare let atlas: any;
+
+const AZURE_LINK = 'https://atlas.microsoft.com/';
+let AZURE_JS_URL = `${AZURE_LINK}sdk/javascript/mapcontrol/3/atlas.min.js`;
+let AZURE_CSS_URL = `${AZURE_LINK}/sdk/javascript/mapcontrol/3/atlas.min.css`;
+let CUSTOM_URL;
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:19]: [refactoring]
Cannot we create a single object? Like
```javascript
const URLS = {
    ...
};
```

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    });
+  },
+
+  _loadMapStyles() {
+    return new Promise<void>((resolve) => {
+      ajax.sendRequest({
+        url: AZURE_CSS_URL,
+        dataType: 'text',
+      }).then((css) => {
+        $('<style>').html(css).appendTo($('head'));
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:153]: We should ensure it works well with CSP policies

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+
+  _destroyMarker(marker) {
+    this._map.markers.remove(marker.marker);
+    if (marker.popup) {
+      this._map.popups.remove(marker.popup);
+    }
+    if (marker.handler) {
+      this._map.events.remove(marker.handler);
+    }
+  },
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:355]: kindly test manually that memory is released well when markers/routes are used

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+          const lineLayer = new atlas.layer.LineLayer(dataSource, null, lineLayerConfig);
+
+          this._map.sources.add(dataSource);
+          this._map.layers.add(lineLayer);
+
+          const bounds = atlas.data.BoundingBox.fromPositions(routeCoordinates);
+          if (this._option('autoAdjust')) {
+            this._map.setCamera({
+              bounds,
+              padding: 50,
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:393]: How have u chosen 50?
Let's move to a constant

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+            });
+          }
+
+          resolve({
+            instance: { dataSource, lineLayer },
+            northEast: [bounds[2], bounds[3]],
+            southWest: [bounds[0], bounds[1]],
+          });
+        }
```
Comment (ksercs) [inline packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:402]: [refactoring]
Cannot we split this code? For example implement separate `_createLineLayer` method, or maybe smth else
Not it's difficult to read

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js (hunk 1 above)
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js]: done

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js (hunk 2 above)
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js]: updated everywhere except for cases with 2-3 defferred

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js (hunk 3 above)
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js (hunk 4 above)
Comment (Zedwag) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/mapParts/azureTests.js:952]: added

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts (hunk 5 above)
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:19]: i tried to add object and in my opinion it looks even worse

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts (hunk 8 above)
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:393]: it's a recommended value from some samples, moved to contant

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts (hunk 9 above)
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:402]: I removed logic related to stroke dash adding for walking route. Now there is no excess if and no separate LineLayer config, I think there is no need to move it to a separate function now

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts (hunk 7 above)
Comment (Zedwag) [reply packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:355]: tested, didn't find any leaks

File: packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts (hunk 5 above)
Comment (ksercs) [reply packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts:19]: We at least can make naming for `CUSTOM_URL` mode precise. But okay, it's up to you

**Review by ksercs — APPROVED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Pull request merged**